<a href="https://colab.research.google.com/github/smaciasg/DataScience2022-3/blob/main/Data_Cleaning_Overview.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Cleaning



> "Garbage in, garbage out."  

Let's clean up the garbage!

In this notebook we will:

1. Examine the head of a dataframe to get a sense of the data
2. Check for duplicate rows and remove them as needed.
3. Consider which columns have information useful for analysis or machine learning,
4. Examine the data types of columns and change them as needed.
5. Check for inconsistent values in categorical columns.



In [ ]:
# Import Pandas
import pandas as pd

# The Data

We will be using a slightly altered version of the Titanic dataset, which is a common practice dataset for learning data science.

The dataset contains information about passengers on The Titanic as well a column describing whether they survived the accident or not.  The ultimate goal of most predictions on this dataset is to predict who will survive based on the other information.  Our goal will be to clean the data and make it ready for analysis.

Each row contains data about one passenger.

# Data Dictionary

Sometimes your data will come with a dictionary that helps you understand what the columns mean and what data types should be in them.  This is useful for cleaning the data.

| Column   | Description                                | Value Keys                                     | Type  |
|----------|--------------------------------------------|------------------------------------------------|-------|
| survival | Survival                                   | 0 = No, 1 = Yes                                | int   |
| pclass   | Ticket class                               | 1 = 1st, 2 = 2nd, 3 = 3rd                      | int   |
| sex      | Sex                                        |                                                | string   |
| Age      | Age in years                               |                                                | int   |
| sibsp    | # of siblings / spouses aboard the Titanic |                                                | int   |
| parch    | # of parents / children aboard the Titanic |                                                | int   |
| ticket   | Ticket number                              |                                                | string   |
| fare     | Passenger fare                             |                                                | float |
| cabin    | Cabin number                               |                                                | string   |
| embarked | Port of Embarkation                        | C = Cherbourg, Q = Queenstown, S = Southampton | string   |


In [ ]:
# Load the Data
df = pd.read_csv('https://docs.google.com/spreadsheets/d/e/2PACX-1vQ7Xta6T-hqBFfQTUMNSVdl03z74KlyecepwOK0Y4s_zv0CG6L8NrVP8tlBsnV2tI4W_cHE5NB9cduU/pub?output=csv')

## df.head()
`df.head()` is one of our most important methods.  It lets us examine the first 5 rows of our dataframe and as many columns as our maximum display columns configuration will allow.

Look below and try to decide which columns will be useful for analysis and predictions, and which columns probably do not contain any information that would be relevant to who might survive the catastrophic sinking of The Titanic.

In [ ]:
# Examine the first 5 rows
df.head()

Unnamed: 0  PassengerId  Survived  Pclass  \
0           0            1         0       3   
1           1            2         1       1   
2           2            3         1       3   
3           3            4         1       1   
4           4            5         0       3   

                                                Name     Sex   Age SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0     1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0     1   
2                             Heikkinen, Miss. Laina  female  26.0     0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0     1   
4                           Allen, Mr. William Henry    male  35.0     0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

# Remove Duplicates

We don't want any duplicate entries in our data.  This will skew our analysis and confuse our predictive models.  Duplicate entries are quite common in data that has not been cleaned.

We can use `df.duplicated()` to show whether rows are duplicates.

In [ ]:
# Check for duplicates
df.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
901     True
902     True
903     True
904     True
905     True
Length: 906, dtype: bool

Well, we don't want to have to look row by row, so let's use `.sum()` to add up all of the `True` values.  When using `.sum()`, a `True` value will counted as 1. A `False` value will not be counted (or counted as 0).

In [ ]:
# Count the duplicates
df.duplicated().sum()

15

We have 15 duplicated rows.  We can simply remove those with df.drop_duplicates().

In [ ]:
# Remove duplicates
df = df.drop_duplicates()
df.duplicated().sum()

0

# Remove Unnecessary Columns

The `Unnamed: 0` column is probably a duplicate of the index.  You'll see this sometimes as Pandas tries to assign a new index when one already exists in the data.  We will drop this column.

The `PassengerID` is a numbering of passengers.  Most likely the passenger number will not affect whether the passenger survived. Each value in this column is unique, is not ordered in any way, and does not describe any quantitative feature of the passenger.  We will drop this column.

The `Name` column is the passenger's name.  There is probably little or no correlation between a passenger's name and whether they survived.  Like `PassengerID` they are probably unique to each row and probably won't be useful for analysis or prediction.  We will drop this column.

In [ ]:
# Remove unnecessary columns
df = df.drop(columns=['Unnamed: 0', 'PassengerId', 'Name'])
df.head()

Survived  Pclass     Sex   Age SibSp  Parch            Ticket     Fare  \
0         0       3    male  22.0     1      0         A/5 21171   7.2500   
1         1       1  female  38.0     1      0          PC 17599  71.2833   
2         1       3  female  26.0     0      0  STON/O2. 3101282   7.9250   
3         1       1  female  35.0     1      0            113803  53.1000   
4         0       3    male  35.0     0      0            373450   8.0500   

  Cabin Embarked  
0   NaN        S  
1   C85        C  
2   NaN        S  
3  C123        S  
4   NaN        S

## Further study of columns with df.nunique()

`.nunique()` will tell us how many unique values are in each column.  This can give us a clue about the liklihood of finding a relationship between the column value and other data.

In [ ]:
# Examine the number of unique values in columns
df.nunique()

Survived      2
Pclass        3
Sex           3
Age          88
SibSp         8
Parch         7
Ticket      681
Fare        248
Cabin       147
Embarked      3
dtype: int64

`Ticket` is a category, but there are almost as many different unique values as there are rows in the dataframe.  If the Ticket data is different for most every passenger and it's not a number describing some quality of the passenger, like 'Age', then it probably won't be very useful.  We will drop that, too.

## Notice
We are using judgement calls here.  Data scientists often must use their judgement, and it's often useful to consult a subject matter expert to double check our judgement when possible.

In [ ]:
# Drop the Ticket column
df = df.drop(columns='Ticket')

# Correct Data Types

## .info() 
`df.info()` is a very useful method.  It tells us a lot about our data.  What information can you gain from looking at the output below?

In [ ]:
# Check columns, rows, data types, and missing values
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    object 
 3   Age       714 non-null    float64
 4   SibSp     891 non-null    object 
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
 7   Cabin     204 non-null    object 
 8   Embarked  889 non-null    object 
dtypes: float64(2), int64(3), object(4)
memory usage: 69.6+ KB


We see that there are 906 entries (rows), 9 columns remaining, and we can see how many values there are in each column and what their datatype is.  We can compare these datatypes to the data dictionary, or to our own understanding of what the data represents.

Notice that `SibSp` is an object data type.  That column represents the number of siblings and spouses a passenger has aboard with them, so it should be an integer.  Let's examine that more closely.

## Correcting Datatypes

We can use `df[column].astype(type)` to change the datatype of a column. For instance we want to change `df['SibSp']` from an object to an integer.

In [ ]:
# Change the SibSp column from an object to an integer
df['SibSp'] = df['SibSp'].astype(int)

ValueError: ignored

## Errors Changing Types

Oh!  We got an error!  If we look carefully at the bottom of the error it says: `ValueError: invalid literal for int() with base 10: 'one'`

It says it can't change the string 'one' to an integer.  That means one of the values is 'one' instead of 1!

We can use `.replace` to change those, even if there are more than one of them.

In [ ]:
# Replace the bad value in SibSp and try again to change the type
df['SibSp'] = df['SibSp'].replace('one', 1)
df['SibSp'] = df['SibSp'].astype(int)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    object 
 3   Age       714 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
 7   Cabin     204 non-null    object 
 8   Embarked  889 non-null    object 
dtypes: float64(2), int64(4), object(3)
memory usage: 69.6+ KB


`df['Age']` should also be an integer, but it is shown as a float.  This usually doesn't matter very much, as machine learning and most visualizations will display ints and floats the same.  But, what happens if we try to change the datatype?

In [ ]:
df['Age'] = df['Age'].astype(int)

IntCastingNaNError: ignored

The `df['Age']` contains NA values, or missing values.  Missing values cannot be converted to integer columns.  When you see a column that should be an integer that is a float, often times this is the reason.

We will deal with the missing values in the 'Age' column in another lesson, so we will leave it as float type data for now.

# Check for Inconsistencies in Categorical Values

Another common problem with dirty data will be categories that are not uniform.  This can include misspellings, abbreviations or typos.

We can use `.value_counts()` again to examine our `object` type columns.  If we do this with numeric columns our outputs can often be too long to read easily.  It works best with categorical variables or columns.

In [ ]:
# Check the unique values and counts of the 'Sex' column
df['Sex'].value_counts()

male      574
female    313
F           4
Name: Sex, dtype: int64

We see that there are both 'female' and 'F' values.  We can assume these are meant to be the same, so we will replace 'F' with 'female' so that our categories are uniform.

## Ensure Consistency in Caegorical Values

In [ ]:
# Standardize the values in the Sex column
df['Sex'] = df['Sex'].replace('F', 'female')
df['Sex'].value_counts()

male      574
female    317
Name: Sex, dtype: int64

# Missing Values

Missing values must always be identified and addressed. Dealing with missing values can be complex, so it will be explored in another lesson. For now, we will start by identify them.

In [ ]:
df.isna().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Cabin       687
Embarked      2
dtype: int64


`Age` has some missing (null) values, `Cabin` has a lot of missing values, and `Embarked` has just 2 missing values.  We will learn how to deal with missing values in another lesson.

# Summary

In this lesson you learned how to examine data using `.head()`, `.info()`, `.nunique()` and `.value_counts()`.  You learned how to identify and remove duplicates, change incorrect datatypes, and how to ensure consistency in numeric and categorical columns.